### PR TITLE
add filter policy interface && add set_cache_index_and_filter_blocks

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
1)add set_filter_policy for BlockBasedOptions; 2)implement create bloom filter.

### CODE REVIEW

**Comment by ngaut** [comment]
PTAL @BusyJay @siddontang

**Comment by siddontang** [comment]
Any test?

**Comment by zhangjinpeng1987** [comment]
PTAL @siddontang @BusyJay @ngaut

**Comment by siddontang** [comment]
LGTM

File: src/rocksdb_options.rs [hunk 1]
```diff
@@ -90,6 +114,20 @@ impl BlockBasedOptions {
             rocksdb_ffi::rocksdb_block_based_options_set_block_cache(self.inner, cache);
         }
     }
+
+    pub fn set_filter_policy(&mut self, filter: &FilterPolicy) {
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: What if the filter is dropped but the option hasn't?

Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: This implement is refer to "set_block_based_table_factory", is there need lifetime?

Comment (BusyJay) [inline src/rocksdb_options.rs]: It depends on whether rocksdb copy the policy when setting. Maybe we can construct some tests to verify or lookup the source code.

Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: OK, I will refer to RocksDB's code.

Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: You are right, 'set_block_based_table_factory' can work in this way, cause rocksdb copy the 'block_based_table_options_t', but set_filter_policy don't.

Comment (BusyJay) [inline src/rocksdb_options.rs]: You can introduce lifetime to get around it. Or just simplify move the policy and keep a reference inside option.

File: src/rocksdb_options.rs [hunk 2]
```diff
@@ -28,6 +28,10 @@ pub struct FilterPolicy {
 
 pub struct BlockBasedOptions {
     inner: rocksdb_ffi::DBBlockBasedTableOptions,
+
+    // FilterPolicy must live as long as BlockBasedTableOptions,
```
Comment (BusyJay) [inline src/rocksdb_options.rs]: Should it live as long as DB?

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

File: src/rocksdb_options.rs (hunk 2 above)
Comment (zhangjinpeng1987) [inline src/rocksdb_options.rs]: rocksdb use shared_ptr to hold filter_policy object, so we don't need call rocksdb_filterpolicy_destroy or keep FilterPolicy live as long as DB.

Comment (BusyJay) [inline src/rocksdb_options.rs]: But it will be automatically dropped when filter_policy goes out of scope.

File: librocksdb_sys/src/lib.rs [hunk 3]
```diff
@@ -200,7 +200,6 @@ extern "C" {
     pub fn rocksdb_options_set_disable_auto_compactions(options: DBOptions,
                                                         v: c_int);
     pub fn rocksdb_options_set_report_bg_io_stats(options: DBOptions, v: c_int);
-    pub fn rocksdb_filterpolicy_destroy(filter_policy: DBFilterPolicy);
```
Comment (BusyJay) [inline librocksdb_sys/src/lib.rs]: If we create the policy and not use it, it will be leaked.

**Comment by zhangjinpeng1987** [comment]
@BusyJay PTAL

File: librocksdb_sys/src/lib.rs (hunk 3 above)
Comment (zhangjinpeng1987) [inline librocksdb_sys/src/lib.rs]: As we can see, rocksdb use shared_ptr to handle filter policy's lifetime, and we don't need free it manually.
```
BlockBasedTableOptions table_options;
table_options.hash_index_allow_collision = false;
table_options.filter_policy.reset(
           NewBloomFilterPolicy(10, use_block_based_builder_));
options_.table_factory.reset(NewBlockBasedTableFactory(table_options));
```

**Comment by BusyJay** [comment]
LGTM

**Comment by zhangjinpeng1987** [comment]
@ngaut @disksing PTAL

**Comment by disksing** [comment]
LGTM

**Pull request merged**